// relay-runtime/lib/RelayDebugger.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @providesModule RelayDebugger
 * 
 * @format
 */

'use strict';

var _classCallCheck3 = _interopRequireDefault(require('babel-runtime/helpers/classCallCheck'));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var RelayDebugger = function () {
  function RelayDebugger() {
    (0, _classCallCheck3['default'])(this, RelayDebugger);

    this._idCounter = 1;
    this._envDebuggers = new Map();
  }

  RelayDebugger.prototype.registerEnvironment = function registerEnvironment(env) {
    var idString = 'RelayModernEnvironment' + this._idCounter++;
    this._envDebuggers.set(idString, new EnvironmentDebugger(env));
    return idString;
  };

  RelayDebugger.prototype.getEnvironmentDebugger = function getEnvironmentDebugger(id) {
    var envDebugger = this._envDebuggers.get(id);
    if (!envDebugger) {
      throw new Error('No registered environment: ' + id);
    }

    return envDebugger;
  };

  RelayDebugger.prototype.getRegisteredEnvironmentIds = function getRegisteredEnvironmentIds() {
    return Array.from(this._envDebuggers.keys());
  };

  return RelayDebugger;
}();

var EnvironmentDebugger = function () {
  function EnvironmentDebugger(environment) {
    (0, _classCallCheck3['default'])(this, EnvironmentDebugger);

    this._environment = environment;
  }

  EnvironmentDebugger.prototype.getEnvironment = function getEnvironment() {
    return this._environment;
  };

  EnvironmentDebugger.prototype.getMatchingRecords = function getMatchingRecords(matchStr, matchType) {
    var inspector = require('./RelayRecordSourceInspector').getForEnvironment(this._environment);

    function isMatching(record) {
      if (matchType === 'idtype') {
        return record.id.includes(matchStr) || !!record.type && record.type.includes(matchStr);
      }
      if (matchType === 'id') {
        return record.id.includes(matchStr);
      }
      if (matchType === 'type') {
        return !!record.type && record.type.includes(matchStr);
      }
      if (matchType === 'predicate') {
        var recordInspector = inspector.get(record.id);
        var fields = recordInspector && recordInspector.inspect();
        if (typeof fields === 'object' && fields !== null) {
          throw new Error('Not implemented');
        }
        return false;
      }
      throw new Error('Unknown match type: ' + matchType);
    }

    return inspector.getRecords().filter(isMatching);
  };

  EnvironmentDebugger.prototype.getRecord = function getRecord(id) {
    var inspector = require('./RelayRecordSourceInspector').getForEnvironment(this._environment);
    var recordInspector = inspector.get(id);
    return recordInspector && recordInspector.inspect();
  };

  return EnvironmentDebugger;
}();

module.exports = RelayDebugger;
